Add margin-trim to CSS parser.

sammygill opened the pull request:
#### e603e43fbb226e2dd51867afc86e6c0a879c800e
<pre>
Add margin-trim to CSS parser.
<a href="https://bugs.webkit.org/show_bug.cgi?id=249205">https://bugs.webkit.org/show_bug.cgi?id=249205</a>
rdar://103285644

Reviewed by Tim Nguyen.

Creates a feature flag for the margin-trim CSS property and also
adds it to the CSS parser behind this flag. When the flag is enabled the
property can be set according to the grammar in the spec.

This property has the following grammar:
none | block | inline | [ block-start || inline-start || block-end || inline-end ]

Spec reference: <a href="https://www.w3.org/TR/css-box-4/#margin-trim">https://www.w3.org/TR/css-box-4/#margin-trim</a>

* LayoutTests/imported/w3c/web-platform-tests/css/css-box/parsing/margin-trim-computed-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/parsing/margin-trim-computed.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/parsing/margin-trim-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/parsing/margin-trim.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* Source/WTF/Scripts/Preferences/WebPreferencesExperimental.yaml:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle):
* Source/WebCore/css/parser/CSSParserContext.cpp:
* Source/WebCore/css/parser/CSSParserContext.h:
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeMarginTrim):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
* Source/WebCore/rendering/style/RenderStyle.h:
(WebCore::RenderStyle::marginTrim const):
(WebCore::RenderStyle::setMarginTrim):
(WebCore::RenderStyle::initialMarginTrim):
* Source/WebCore/rendering/style/RenderStyleConstants.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/rendering/style/RenderStyleConstants.h:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
(WebCore::StyleRareNonInheritedData::StyleRareNonInheritedData):
(WebCore::StyleRareNonInheritedData::operator== const):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertMarginTrim):

Canonical link: <a href="https://commits.webkit.org/257960@main">https://commits.webkit.org/257960@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e43d22f400b9a161e2c72ca063c85562864e7a37

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/100444 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/9604 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/33506 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/109800 "Built successfully") | [❌ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/170047 "Hash e43d22f4 for PR 7626 does not build (failure)") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/104436 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/10516 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/85/builds/216 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/92867 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/107633 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/106223 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/7959 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/91227 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/34610 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/89875 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/22616 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/77572 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/90985 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/3340 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/24135 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/87006 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/801 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/3332 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/85/builds/216 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/29123 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/9453 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/43625 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/89888 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5454 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/5147 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/20094 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/parser/CSSParserContext.cpp:
```diff
@@ -89,6 +89,7 @@ CSSParserContext::CSSParserContext(const Document& document, const URL& sheetBas
     , gradientInterpolationColorSpacesEnabled { document.settings().cssGradientInterpolationColorSpacesEnabled() }
     , subgridEnabled { document.settings().subgridEnabled() }
     , masonryEnabled { document.settings().masonryEnabled() }
+    , marginTrimEnabled { document.settings().cssMarginTrimEnabled() }
```
This shouldn't be needed since you're using `"settings-flag": "cssMarginTrimEnabled"` in `CSSProperties.json`

sammygill replied:
Removed

nt1m commented on Source/WebCore/css/parser/CSSParserContext.h:
```diff
@@ -96,6 +96,7 @@ struct CSSParserContext {
     bool gradientInterpolationColorSpacesEnabled { false };
     bool subgridEnabled { false };
     bool masonryEnabled { false };
+    bool marginTrimEnabled { false };
```
ditto

sammygill replied:
Removed

nt1m commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -5591,6 +5591,34 @@ RefPtr<CSSValue> consumeMarginSide(CSSParserTokenRange& range, CSSPropertyID cur
+RefPtr<CSSValue> consumeMarginTrim(CSSParserTokenRange& range, const CSSParserContext& context)
+{
+    if (!context.marginTrimEnabled)
+        return nullptr;
```
ditto

sammygill replied:
Removed

nt1m commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -5591,6 +5591,34 @@ RefPtr<CSSValue> consumeMarginSide(CSSParserTokenRange& range, CSSPropertyID cur
+    if (!context.marginTrimEnabled)
+        return nullptr;
+    auto firstValue = range.peek().id();
+    if (firstValue == CSSValueBlock || firstValue == CSSValueInline  || firstValue == CSSValueNone)
```
```suggestion
    if (firstValue == CSSValueBlock || firstValue == CSSValueInline || firstValue == CSSValueNone)
```

sammygill replied:
Fixed nice catch!

nt1m commented on Source/WebCore/rendering/style/RenderStyle.h:
```diff
@@ -1684,6 +1688,7 @@ class RenderStyle {
     static Length initialOffset() { return Length(); }
     static Length initialRadius() { return Length(); }
     static Length initialMargin() { return Length(LengthType::Fixed); }
+    static OptionSet<MarginTrimType> initialMarginTrim() { return OptionSet<MarginTrimType> { }; }
```
`return { };` ?

sammygill replied:
Changed

nt1m commented on Source/WebCore/css/CSSProperties.json:
```diff
@@ -3700,6 +3700,18 @@
+                "settings-flag": "cssMarginTrimEnabled",
+                "converter": "MarginTrim",
+                "custom-parser": true,
+                "parser-requires-context": true
```
@weinig Is the new parser good enough for this use-case?

nt1m commented on Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
```diff
@@ -249,6 +249,8 @@ class StyleRareNonInheritedData : public RefCounted<StyleRareNonInheritedData> {
     unsigned leadingTrim : 2; // LeadingTrim
 
+    OptionSet<MarginTrimType> marginTrim;
+
```
Can you move this up so all the `unsigned` bits can be packed together?

sammygill replied:
Yup! Done

nt1m commented on Source/WebCore/css/CSSValueKeywords.in:
```diff
@@ -732,6 +732,10 @@ column-reverse
 wrap-reverse
 
+// CSS_PROP_MARGIN_TRIM
+block-start
+block-end
```
For consistency with other props:
```suggestion
// block
block-start
block-end
// inline
// inline-start
// inline-end
```

sammygill replied:
Changed (inline-start and inline-end were already defined for other properties it seems so I moved them)

nt1m replied:
inline-start and inline-end are also valid values for `float` and `clear`, so they shouldn't be removed from the section above.

You should include them again here but commented out.

sammygill replied:
Done

sammygill committed:
Add margin-trim to CSS parser.
https://bugs.webkit.org/show_bug.cgi?id=249205
rdar://103285644

Reviewed by Tim Nguyen.

Creates a feature flag for the margin-trim CSS property and also
adds it to the CSS parser behind this flag. When the flag is enabled the
property can be set according to the grammar in the spec.

This property has the following grammar:
none | block | inline | [ block-start || inline-start || block-end || inline-end ]

Spec reference: https://www.w3.org/TR/css-box-4/#margin-trim

* LayoutTests/imported/w3c/web-platform-tests/css/css-box/parsing/margin-trim-computed-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/parsing/margin-trim-computed.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/parsing/margin-trim-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/parsing/margin-trim.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* Source/WTF/Scripts/Preferences/WebPreferencesExperimental.yaml:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle):
* Source/WebCore/css/parser/CSSParserContext.cpp:
* Source/WebCore/css/parser/CSSParserContext.h:
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeMarginTrim):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
* Source/WebCore/rendering/style/RenderStyle.h:
(WebCore::RenderStyle::marginTrim const):
(WebCore::RenderStyle::setMarginTrim):
(WebCore::RenderStyle::initialMarginTrim):
* Source/WebCore/rendering/style/RenderStyleConstants.cpp:
(WebCore::operator<<):
* Source/WebCore/rendering/style/RenderStyleConstants.h:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
(WebCore::StyleRareNonInheritedData::StyleRareNonInheritedData):
(WebCore::StyleRareNonInheritedData::operator== const):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertMarginTrim):

Canonical link: https://commits.webkit.org/257960@main
